### PR TITLE
feat: add HalfCheetah compile benchmarking

### PR DESCRIPTION
**Pull request by CarsonBurke**
Add opt-in `torch.compile` support for the HalfCheetah training hot paths and a benchmark script to measure whether it pays off, so compile decisions are driven by measured warm-step speedups instead of guesses.

- Wrap the world-model loss, imagination generation, and imagination learning paths in compile-friendly `nn.Module` callables (new `compile_runtime.py`, shared by the train and benchmark scripts) so `torch.compile` sees stable tensor-only signatures instead of dataclass plumbing.
- Default `compile_mode` to `reduce-overhead`, but only let CUDA Graphs capture the no-grad `imagination_generate` path; the autograd loss wrappers are normalized to `default` / `max-autotune-no-cudagraphs` because graph replay can overwrite AOTAutograd forward intermediates that backward still needs. `torch.compiler.cudagraph_mark_step_begin()` is called before each captured generate call and dream tensors never outlive their iteration.
- Rewrite data-dependent ops on the compiled paths to be graph-safe: a local `masked_mean` shadow that keeps the `torch_einops_utils` semantics without boolean advanced indexing (verified numerically equivalent incl. gradients, broadcast masks, and the complex ECF path), a `masked_fill` PMPO pos/neg loss, tensor `num_advantages`, and `sample_prob` 0/1 short-circuits. The autoregressive action-loss gate is precomputed as a python bool in `__init__` so zero-weight configs still skip the dead branch without a per-step tensor sync.
- `benchmark_halfcheetah_compile.py` compares eager vs compiled default / reduce-overhead / max-autotune arms with per-arm compile-state reset (`torch.compiler.reset` + `gc.collect` + `empty_cache`), peak CUDA memory reporting, and cold/warm accounting that treats the first 3 calls of CUDA Graph paths as cold since capture happens on calls 2-3, not just call 1. Warm comparisons use a per-step composite so arms with different warmup counts stay comparable.
- The benchmark rejects user overrides of pinned kwargs instead of silently ignoring them, caps the train window to rollout length (longer windows are pure padding), and clamps perf-preset batch sizes so compiled shapes stay static across loops instead of recompiling mid-run.
- Timing instrumentation (with its device-wide syncs) now only activates for `track_compile_performance=True` runs such as the benchmark; plain `--compile=True` training runs keep async execution.

Validation on an RTX 5090 (tiny batch-size-1 smoke shape, 2 loops, no-prompt PPO): reduce-overhead with generation CUDA Graphs is the best arm — 12.9x warm per-step speedup vs eager (default: 10.6x, max-autotune: 6.7x), lowest cold compile+capture time (46.8s vs 66.9s / 180.7s), break-even after ~267 warm steps, and eager-level peak memory (105.6 MiB allocated vs 180.2 / 173.9 MiB for the other arms), which supports it as the default. Also ran the CPU eager smoke path end-to-end, targeted `pytest` selections over the touched model code, and a subagent review pass (3 reviewers) whose findings are addressed in the final commits.

Closes #13

### CODE REVIEW

File: train_halfcheetah_imagination_rl.py [hunk 1]
```diff
@@ -919,6 +1356,124 @@ def resolve_log_dir(log_dir: str, checkpoint_path: str | None):
...
+        track_compile_performance = True,
+        compile_fullgraph = True,
+        compile_dynamic = False,
+        require_cuda = benchmark_require_cuda,
+        prob_shortcut_train = 1.,
+    )
+
+    results = []
+
+    for name, compile_enabled in (("eager", False), ("compiled", True)):
```
Comment (witherhoard99) [inline train_halfcheetah_imagination_rl.py]: Would be nice to add something like 
```python
for name, compile_enabled, compile_mode in (("eager", False, None), ("compiled_default", True, "default"), ("compiled_reduceoverhead", True, "reduce-overhead"), ("compiled_maxautotune", True, "max-autotune")):
```

and then pass that along
